### PR TITLE
0.5.1: additional supported operations

### PR DESCRIPTION
**Pull request by mlodi-hqs**
(Latest update 23/2)
Adds support for (almost) all qoqo Operatiions. The backend does not rely on qelib1.inc as all gate definitions are prepended to the QASM file.
As for the moment, this adds support for the following operations:
- PragmaSleep (ALLOWED)
- PragmaGlobalPhase (ALLOWED)
- PragmaStopParallelBlock (ALLOWED)
- PragmaStopDecompositionBlock (ALLOWED)
- PragmaStartDecompositionBlock (ALLOWED)
- PhaseShiftState1
- InvSqrtPauliX
- VariableMSXX
- ControlledPhaseShift
- SWAP
- PragmaActiveReset
- PragmaConditional
- RotateXY
- XY
- ISwap
- SqrtISwap
- InvSqrtISwap
- FSwap
- PMInteraction
- Fsim
- Qsim
- GivensRotation
- GivensRotationLittleEndian
- PhaseShiftedControlledZ
- PhaseShiftedControlledPhase

Missing (math problems):
- PhaseShiftState0
- Bogoliubov
- ComplexPMInteraction
- RotateAroundSphericalAxis

### CODE REVIEW

File: qoqo_qasm/tests/integration/interface.rs [hunk 1]
```diff
@@ -108,7 +108,6 @@ fn test_qasm_call_operation(operation: Operation, converted: &str) {
 
 /// Test qasm_call_operation and qasm_call_circuit errors
 #[test_case(Operation::from(ISwap::new(0, 1)))]
-#[test_case(Operation::from(ControlledPhaseShift::new(0, 1, CalculatorFloat::from(0.23))))]
 #[test_case(Operation::from(FSwap::new(0, 1)))]
```
Comment (kbarkhqs) [inline qoqo_qasm/tests/integration/interface.rs:111]: How come this has been deleted?

Comment (mlodi-hqs) [reply qoqo_qasm/tests/integration/interface.rs:111]: ControlledPhaseShift has been implemented, so it shouldn't trigger an error

Comment (kbarkhqs) [reply qoqo_qasm/tests/integration/interface.rs:111]: Ah fair, has it been added to the tests that should pass?

File: roqoqo-qasm/src/interface.rs [hunk 2]
```diff
@@ -18,7 +18,15 @@ use roqoqo::Circuit;
 use roqoqo::RoqoqoBackendError;
 
 // Pragma operations that are ignored by backend and do not throw an error
-const ALLOWED_OPERATIONS: &[&str; 2] = &["PragmaSetNumberOfMeasurements", "InputSymbolic"];
+const ALLOWED_OPERATIONS: &[&str; 7] = &[
+    "PragmaSleep",
```
Comment (kbarkhqs) [inline roqoqo-qasm/src/interface.rs:22]: Should these be allowed or should they raise an error?

File: qoqo_qasm/tests/integration/interface.rs (hunk 1 above)
Comment (kbarkhqs) [reply qoqo_qasm/tests/integration/interface.rs:111]: Similarly, have the new gates you added been added too? I can't see them in the test

File: roqoqo-qasm/src/interface.rs [hunk 3]
```diff
@@ -154,6 +193,32 @@ pub fn call_operation(
...
+            Ok(format!("reset {}[{}];", qubit_register_name, op.qubit(),))
+        }
+        Operation::PragmaConditional(op) => {
+            // can't handle multiple operations under if condition
+            let mut ite = op.circuit().iter().peekable();
+            let mut data = "".to_string();
+            while let Some(int_op) = ite.next() {
+                if ite.peek().is_none() {
+                    data.push_str(&format!(
+                        "if({}[{}]==1) {}",
```
Comment (kbarkhqs) [inline roqoqo-qasm/src/interface.rs:206]: Should this end with a semi-colon?

File: roqoqo-qasm/src/interface.rs [hunk 4]
```diff
@@ -124,6 +140,14 @@ pub fn call_operation(
             qubit_register_name,
             op.target()
         )),
+        Operation::VariableMSXX(op) => Ok(format!(
+            "rxx({}) {}[{}],{}[{}]",
```
Comment (kbarkhqs) [inline roqoqo-qasm/src/interface.rs:144]: Should this end with a semi-colon?

File: roqoqo-qasm/src/interface.rs [hunk 5]
```diff
@@ -138,6 +162,21 @@ pub fn call_operation(
...
+        Operation::ControlledPhaseShift(op) => Ok(format!(
+            "cp({}) {}[{}],{}[{}];",
+            op.theta(),
+            qubit_register_name,
+            op.control(),
+            qubit_register_name,
+            op.target()
+        )),
+        Operation::SWAP(op) => Ok(format!(
+            "swap {}[{}],{}[{}]",
```
Comment (kbarkhqs) [inline roqoqo-qasm/src/interface.rs:174]: same here, missing a semi-colon?

File: roqoqo-qasm/tests/integration/interface.rs [hunk 6]
```diff
@@ -36,30 +36,49 @@ fn tmp_create_map() -> HashMap<usize, usize> {
...
 #[test_case(Operation::from(TGate::new(0)), "t q[0];"; "TGate")]
+#[test_case(Operation::from(PhaseShiftState1::new(0, CalculatorFloat::from(PI))), "p(3.141592653589793) q[0];"; "PhaseShiftState1")]
 #[test_case(Operation::from(RotateX::new(0, CalculatorFloat::from(-PI))), "rx(-3.141592653589793) q[0];"; "RotateX")]
 #[test_case(Operation::from(RotateY::new(0, CalculatorFloat::from(-PI))), "ry(-3.141592653589793) q[0];"; "RotateY")]
 #[test_case(Operation::from(RotateZ::new(0, CalculatorFloat::from(-PI))), "rz(-3.141592653589793) q[0];"; "RotateZ")]
-#[test_case(Operation::from(SqrtPauliX::new(0)), "rx(pi/2) q[0];"; "SqrtPauliX")]
+#[test_case(Operation::from(SqrtPauliX::new(0)), "sx q[0];"; "SqrtPauliX")]
+#[test_case(Operation::from(InvSqrtPauliX::new(0)), "sxdg q[0];"; "InvSqrtPauliX")]
 #[test_case(Operation::from(MolmerSorensenXX::new(0, 1)), "rxx(pi/2) q[0],q[1];"; "MolmerSorensenXX")]
 #[test_case(Operation::from(CNOT::new(0, 1)), "cx q[0],q[1];"; "CNOT")]
```
Comment (kbarkhqs) [inline roqoqo-qasm/tests/integration/interface.rs:46]: I think you're missing SWAP and VariableMSXX here

File: roqoqo-qasm/tests/integration/interface.rs [hunk 7]
```diff
@@ -36,30 +36,49 @@ fn tmp_create_map() -> HashMap<usize, usize> {
...
 #[test_case(Operation::from(DefinitionUsize::new("ro".to_string(), 1, true)), "creg ro[1];"; "DefinitionUsize")]
 #[test_case(Operation::from(DefinitionBit::new("ro".to_string(), 1, true)), "creg ro[1];"; "DefinitionBit")]
 #[test_case(Operation::from(DefinitionComplex::new("ro".to_string(), 1, true)), "creg ro[1];"; "DefinitionComplex")]
-#[test_case(Operation::from(InputSymbolic::new("other".to_string(), 0.0)), ""; "InputSymbolic")]
+#[test_case(Operation::from(PragmaSleep::new(vec![0,1], CalculatorFloat::from(0.3))), ""; "PragmaSleep")]
+#[test_case(Operation::from(PragmaGlobalPhase::new(CalculatorFloat::from(0.3))), ""; "PragmaGlobalPhase")]
+#[test_case(Operation::from(PragmaStopDecompositionBlock::new(vec![0,1])), ""; "PragmaStopDecompositionBlock")]
 #[test_case(Operation::from(PragmaSetNumberOfMeasurements::new(20, "ro".to_string())), ""; "PragmaSetNumberOfMeasurements")]
+#[test_case(Operation::from(PragmaStartDecompositionBlock::new(vec![0,1], HashMap::new())), ""; "PragmaStartDecompositionBlock")]
+#[test_case(Operation::from(InputSymbolic::new("other".to_string(), 0.0)), ""; "InputSymbolic")]
```
Comment (kbarkhqs) [inline roqoqo-qasm/tests/integration/interface.rs:62]: I think you're missing some operations from the allowed ones here

**Review by kbarkhqs — APPROVED** [review]

**Review by kbarkhqs — CHANGES_REQUESTED** [review]
Not all operations have been added to the test, and some of the operations in the interface are missing semi-colons

File: roqoqo-qasm/src/interface.rs (hunk 3 above)
Comment (mlodi-hqs) [reply roqoqo-qasm/src/interface.rs:206]: I'm double checking, but the internal call_operation(...) should handle that.

File: roqoqo-qasm/src/interface.rs (hunk 2 above)
Comment (mlodi-hqs) [reply roqoqo-qasm/src/interface.rs:22]: Allowed. In the sense that they should not stop the process nor add anything to the output. They are ignored.

**Review by kbarkhqs — APPROVED** [review]

**Review by kbarkhqs — APPROVED** [review]

**Review by kbarkhqs — APPROVED** [review]

**Review by nfwvogt — APPROVED** [review]
Looks good to me, but we should remove __version__.py then release

**Pull request merged**